python/paddle/distributed/collective.py

```diff
@@ -365,14 +367,60 @@ def _set_custom_gid(gid):
     _custom_gid = gid
 
 
-def new_group(ranks=None, backend=None):
+def _store_barrier(store, global_rank, global_world_size, timeout, group_name):
+    start_barrier_prefix = "StartBarrier/" + group_name + "/"
+    exit_barrier_key = "CanExitBarrier/" + group_name
+    end_barrier_prefix = "EndBarrier/" + group_name + "/"
+    is_master = global_rank == 0
+
+    def _check_keys_ready(wait_keys):
+        start_time = time.time()
+        while len(wait_keys) > 0:
+            time.sleep(0.5)
+            elapse_time = time.time() - start_time
+            if datetime.timedelta(seconds=elapse_time) > timeout:
+                raise RuntimeError(
+                    "Timeout while initializing process group {}."
+                    "Keys {} are not ready sinck rank {} is waiting them."
+                    "Two reason may cause this error:\n 1. The create process group api should be called by all ranks.\n"
+                    " 2. Try to increase the waiting time.\n".format(
+                        group_name, wait_keys, global_rank))
+            wait_keys = list(
+                filter(lambda key: int(store.get(key)) != 1, wait_keys))
+
+    # Step 1: all worker processes send and wait for barrier key from master.
+    if is_master:
+        wait_keys = [
+            start_barrier_prefix + str(rank)
+            for rank in range(1, global_world_size)
+        ]
+        _check_keys_ready(wait_keys)
+        store.add(exit_barrier_key, 1)
+    else:
+        store.add(start_barrier_prefix + str(global_rank), 1)
+        wait_keys = [exit_barrier_key]
+        _check_keys_ready(wait_keys)
+
+    # Step 2: workers need to send ack to master, ensure master exit at last
+    if is_master:
+        wait_keys = [
+            end_barrier_prefix + str(rank)
+            for rank in range(1, global_world_size)
+        ]
+        _check_keys_ready(wait_keys)
+    else:
+        store.add(end_barrier_prefix + str(global_rank), 1)
+
+
+def new_group(ranks=None, backend=None, store_timeout=_default_store_timeout):
```

Comment:
store_timeout=>timeout

Reply:
There will have multiple timeout variables in this api. Store timeout used for judge whether tcp store is timeout. Communication timeout used for judge whether nccl, gloo is hang in task.wait().

Reply:
done